lib/wibox/drawable.lua

```diff
@@ -144,6 +144,48 @@ local function do_redraw(self)
         self._widget_hierarchy:draw(context, cr)
     end
 
+    -- If a `gears.shape` is set and the border_width is greater than 2,
+    -- re-apply the border to fix anti-aliasing.
+    if self._widget_context_skeleton.wibox then
+        local wb = self._widget_context_skeleton.wibox
+
+        local shape = wb._shape
+        local bw = wb.border_width
+
+        -- The "real" size is width+2*bw and height+2*bw. This code only exist
+        -- to draw the inner anti-aliased border just 1px inside of the clip.
+        if shape then
+            local bc = wb.border_color or beautiful.border_normal or beautiful.fg_normal
+
+            cr:save()
+
+            -- First, draw the actual border. This will also include an
+            -- anti-aliased inner border that the clip shape could not provide.
+            cr:move_to(0, 0)
+            cr:translate(-bw, -bw)
+            cr:set_source(color(bc))
+            shape(cr, width+2*bw, height+2*bw)
+
+            -- Add 1px when the border_width is 1 to compensate for the
+            -- outer border anti-alising. Yes, this does steal half a pixel
+            -- that should allow visible content.
+            cr:set_line_width(bw*2 + (bw == 1 and 1 or 0))
+            cr:stroke_preserve()
+
+            -- Finally, clear the outer pixel to create an anti-alised effect
+
+            -- Odd numbers are odds, but this is mostly due to the A1 shape
+            -- bounding own anti-alising.
+            cr:set_line_width(1 + (bw % 2)/2)
+
+            cr:set_operator(cairo.Operator.CLEAR)
+            cr:set_source_rgba(0,0,0,1)
+            cr:stroke()
+
+            cr:restore()
+        end
+    end
```

Comment:
I'd prefer to have this wibox-related code in `wibox.lua`. How about adding an after-draw-callback that the wibox code could use?

Then again, what does this actually do and how/why does it work? You take the bounding shape of the wibox and somehow apply it to the drawing after it was done? I do not really see how this can work or does work. Also, this CLEARs some pixels, which means that we now get some black where previously there was no black (unless a compositing manager is running...).

So... how does this work? Should it perhaps only be done when a compositing manager is running?